### ISSUE TITLE
ByteBuffer should have hex bytes initialiser / write method

### ISSUE THREAD

**Issue by weissi** [apple/swift-nio#2748]
This is pretty much a companion of #2447 / #2475. The aforementioned bugs/patches introduce the ability to get the hex dump from a `ByteBuffer` which is awesome.

Next, I'd like to construct/write into a `ByteBuffer` from a hex dump:
- `ByteBuffer(hexEncodedBytes: "68 65 6c 6c 6f 20 77 6f 72 6c 64 0a")`
- `buffer.writeHexEncodedBytes("68656c6c6f20776f726c640a")`

I'd say it just needs to support the `.plain` format (i.e. 2 hex-digit values with whitespace ignored)

---

Super basic test:

```
var buffer = ByteBuffer(hexEncodedBytes: "68 65 6c 6c 6f 20 77 6f 72 6c 64 0a")
buffer.writeHexEncodedBytes("68656c6c6f20776f726c64")
buffer.writeHexEncodedBytes("")
buffer.writeHexEncodedBytes("0a")
XCTAssertEqual(ByteBuffer(string: "hello world\nhello world\n"), buffer)
```

another one:

```
var allBytes = ByteBuffer()
for x in UInt8.min ... UInt8.max {
    buffer.writeInteger(x)
}

let allBytesHex = allBytes.hexDump(format: .plain)
let allBytesDecoded = ByteBuffer(hexEncodedBytes: allBytesHex)

XCTAssertEqual(allBytes, allBytesDecoded)
```

**Comment by ali-ahsan-ali** [comment]
I have taken a look and would like to pick this up after https://github.com/apple/swift-nio/issues/2734.

**Comment by weissi** [comment]
Of course this should also parse the `.compat` format suggested in #2825

**Comment by weissi** [comment]
> Happy if someone else is eager to give it a go. I have started looking at it but no real progress. It feels overly ambitious to try and start working on a second PR when my first has taken quite a while.

I did re-request review on your other PR. Remember to click the  :recycle  (re-request) review once you addressed the changed suggested/requested. I'm sure @FranzBusch & @Lukasa just missed that you pushed updates to your existing PR.

Separately, if you're up for it, you can totally have multiple PRs open at the same time, shouldn't slow things down.

**Comment by ali-ahsan-ali** [comment]
The "good for newcomers" in this repo are the hardest thing every 😅 (which I VERY much enjoy).

Experimenting with the below and seeing if writing that to the buffer is the right outcome for writing a hex encoded string to buffer. 

```
extension String {
    var hexData: some Sequence<UInt8> {
        sequence(state: self, next: { remainder in
            guard remainder.count >= 2 else { return nil }
            let nextTwo = remainder.prefix(2)
            remainder.removeFirst(2)
            return UInt8(nextTwo, radix: 16) // compatible
        })
    }
}
```

TLDR: Thanks for the push. Working on it!

**Comment by weissi** [comment]
> The "good for newcomers" in this repo are the hardest thing every 😅 (which I VERY much enjoy).
> 
> Experimenting with the below and seeing if writing that to the buffer is the right outcome for writing a hex encoded string to buffer.
> 
> ```
> extension String {
>     var hexData: some Sequence<UInt8> {
>         sequence(state: self, next: { remainder in
>             guard remainder.count >= 2 else { return nil }
>             let nextTwo = remainder.prefix(2)
>             remainder.removeFirst(2)
>             return UInt8(nextTwo, radix: 16) // compatible
>         })
>     }
> }
> ```
> 
> TLDR: Thanks for the push. Working on it!

Awesome, FWIW, you can use

```
extension String {
   var hexData: some Sequence<UInt8> {
       return ByteBuffer(string: .self).hexDump(format: .plain).utf8
   }
}
```

to get a sequence of `UInt8`s. `String`'s `.utf8` gives you a `UTF8View` which is a `Sequence<UInt8>`

**Comment by ali-ahsan-ali** [comment]
@weissi please take a look at the PR when you fidn the time

**Comment by ali-ahsan-ali** [comment]
I think this can be closed now

**Comment by ali-ahsan-ali** [comment]
Can this be closed?
